proof/crefine/ARM/Fastpath_Equiv.thy

```diff
@@ -417,48 +410,33 @@ lemma schedule_rewrite_ct_not_runnable':
             (do setSchedulerAction ResumeCurrentThread; switchToThread t od)"
   supply subst_all [simp del]
   apply (simp add: schedule_def)
-  apply (rule monadic_rewrite_imp)
-   apply (rule monadic_rewrite_trans)
-    apply (rule monadic_rewrite_bind_tail)
-     apply (rule monadic_rewrite_bind_tail)
-      apply (rule_tac P="action = SwitchToThread t" in monadic_rewrite_gen_asm, simp)
-      apply (rule monadic_rewrite_bind_tail)
-       apply (rule_tac P="\<not> wasRunnable \<and> action = SwitchToThread t"
-              in monadic_rewrite_gen_asm,simp)
-       apply (rule monadic_rewrite_bind_tail, rename_tac idleThread)
-        apply (rule monadic_rewrite_bind_tail, rename_tac targetPrio)
-         apply (rule monadic_rewrite_bind_tail, rename_tac curPrio)
-          apply (rule monadic_rewrite_bind_tail, rename_tac fastfail)
-           apply (rule monadic_rewrite_bind_tail, rename_tac curDom)
-            apply (rule monadic_rewrite_bind_tail, rename_tac highest)
-             apply (rule_tac P="\<not> (fastfail \<and> \<not> highest)" in monadic_rewrite_gen_asm, simp only:)
-             apply simp
-             apply (rule monadic_rewrite_refl)
-            apply (wpsimp wp: hoare_vcg_imp_lift)
-            apply (simp add: isHighestPrio_def')
-            apply wp+
-          apply (wp hoare_vcg_disj_lift)
-           apply (wp scheduleSwitchThreadFastfail_False_wp)
-          apply (wp hoare_vcg_disj_lift threadGet_wp'' | simp add: comp_def)+
-   (* remove no-ops, somewhat by magic *)
-   apply (rule monadic_rewrite_symb_exec_l'_TT, solves wp,
-          wpsimp wp: empty_fail_isRunnable simp: isHighestPrio_def')+
-            apply (simp add: setSchedulerAction_def)
-            apply (subst oblivious_modify_swap[symmetric], rule oblivious_switchToThread_schact)
-            apply (rule monadic_rewrite_refl)
-           apply wp+
-  apply (clarsimp simp: ct_in_state'_def)
-  apply (strengthen not_pred_tcb_at'_strengthen, simp)
-  supply word_neq_0_conv[simp del]
+  (* switching to t *)
+  apply (monadic_rewrite_l sched_act_SwitchToThread_rewrite[where t=t])
+   (* not wasRunnable, skip enqueue *)
+   apply (simp add: when_def)
+   apply (monadic_rewrite_l monadic_rewrite_if_l_False)
+   (* fastpath: \<not> (fastfail \<and> \<not> highest) *)
+   apply (monadic_rewrite_l monadic_rewrite_if_l_False
+            \<open>wpsimp simp: isHighestPrio_def'
+                    wp: hoare_vcg_imp_lift hoare_vcg_disj_lift threadGet_wp''
+                        scheduleSwitchThreadFastfail_False_wp\<close>)
+   (* fastpath: no scheduleChooseNewThread *)
+   apply (monadic_rewrite_l monadic_rewrite_if_l_False
+            \<open>wpsimp simp: isHighestPrio_def'
+                    wp: hoare_vcg_imp_lift hoare_vcg_disj_lift threadGet_wp''
+                        scheduleSwitchThreadFastfail_False_wp\<close>)
+   (* remove no-ops *)
+   apply (repeat 10 monadic_rewrite_symb_exec_l) (* until switchToThread *)
```

Comment:
I've seen `repeat n blah` a few times now. I'm not sure `repeat_unless` is always the best way to go, but would it be possible to repeat something until a certain term is seen (`switchToThread` in this case)?

Reply:
Definitely possible, but you have to match on the whole conclusion pattern, something like using `repeat_unless` with `has_concl` on `monadic_rewrite ?F ?E ?P (switchToThread ?t >>= ?c1) (?c2)`. Do you think it's worth it?

Reply:
Ah, probably not worth it, then